[Clang-Tidy] Handle nested-name-specifier in "llvm-prefer-isa-or-dyn-cast-in-conditionals"

zwuis opened the pull request:
Use `declRefExpr` matcher to match callee so that we can get the `SourceRange` of the identifier of the callee for replacement.

Drive-by changes:

- Use `hasConditionVariableStatement` matcher to handle `if` statements with init-statement.
- Support `for` loops.

Fixes #154790

zwuis committed:
Handle nested-name-specifier in "llvm-prefer-isa-or-dyn-cast-in-conditionals"

vbvictor commented:
Thank you for the improvements of the check! Did you try running on LLVM code to check for any miscodegen/FP? https://clang.llvm.org/extra/clang-tidy/#running-clang-tidy-in-parallel

EugeneZelenko commented on clang-tools-extra/clang-tidy/llvm/PreferIsaOrDynCastInConditionalsCheck.cpp:
```diff
@@ -22,105 +23,95 @@ AST_MATCHER(Expr, isMacroID) { return Node.getExprLoc().isMacroID(); }
-    StringRef ArgString(Lexer::getSourceText(
-        CharSourceRange::getTokenRange(Arg->getSourceRange()),
-        *Result.SourceManager, getLangOpts()));
+    StringRef LHSString = GetText(LHS->getSourceRange());
+
+    StringRef ArgString = GetText(Arg->getSourceRange());
```
```suggestion
    const StringRef LHSString = GetText(LHS->getSourceRange());
    const StringRef ArgString = GetText(Arg->getSourceRange());
```

EugeneZelenko commented on clang-tools-extra/clang-tidy/llvm/PreferIsaOrDynCastInConditionalsCheck.cpp:
```diff
@@ -22,105 +23,95 @@ AST_MATCHER(Expr, isMacroID) { return Node.getExprLoc().isMacroID(); }
-
-    std::string Replacement("isa_and_nonnull");
-    Replacement += RHSString.substr(Func->getName().size());
+    std::string Replacement = llvm::formatv(
```
```suggestion
    const std::string Replacement = llvm::formatv(
```

zwuis committed:
Remove empty line, add `const`, and make known string inline

zwuis committed:
Use `assert` instead of branch

flovent commented on clang-tools-extra/test/clang-tidy/checkers/llvm/prefer-isa-or-dyn-cast-in-conditionals.cpp:
```diff
@@ -76,6 +136,16 @@ bool foo(Y *y, Z *z) {
+  if (z->bar() && cast_if_present<Y>(z->bar()))
+    return true;
+  // CHECK-MESSAGES: :[[@LINE-2]]:7: warning: isa_and_nonnull<> is preferred
+  // CHECK-FIXES: if (isa_and_nonnull<Y>(z->bar()))
```
Maybe we should suggest `isa_and_present` here? instead of hard coded `isa_and_nonnull`.

flovent replied:
>  Keep _present and _or_null as written.

I think that's reasonable, keeping their original style.

> Replace x && isa<Y>(x) with what configuration files choose (isa_and_nonnull or isa_and_present).

What default option should we use here, i think we don't have cast APIs with `present` postfix when this check was added to clang-tidy.

zwuis replied:
> What default option should we use here

I have no idea. 🤦 We still don't reach a consense on preferring `isa_and_nonnull` or `isa_and_present`.

https://discourse.llvm.org/t/psa-swapping-out-or-null-with-if-present/65018

HerrCai0907 reviewed (approved):
LGTM.
since you have already changing the matcher a lot. could you give more meaningful name for each locals. e.g. Any, Condition, ...

zwuis committed:
Use more meaningful variable names, and revert supporting `cast_if_present` and `dyn_cast_if_present`
Let's wait until reaching a consense on preferring `isa_and_nonnull` or `isa_and_present`

zwuis committed:
Merge branch 'main' into fix#154790

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -249,6 +249,16 @@ Changes in existing checks
   <clang-tidy/checks/readability/uppercase-literal-suffix>` check to recognize
   literal suffixes added in C++23 and C23.
 
+- Improved :doc:`llvm-prefer-isa-or-dyn-cast-in-conditionals
```
Please keep alphabetical order (by check name) in this list.

zwuis committed:
Keep alphabetical order

vbvictor reviewed (approved):
LGTM

zwuis committed:
Add `const` and fix case
